[Sema] Fix `Differentiable` derived conformances crasher.

dan-zheng opened the pull request:
Previously, `getUnderlyingAllDiffableVariables` called `ProtocolConformance::getConcrete()`
unconditionally, assuming that the `ProtocolConformance` is concrete. However, the
conformance may be abstract (for generic type parameter types, for instance).

Check whether conformance is concrete before calling `getConcrete`. Tidy up related code.

Enable `-verify-ignore-unknown` for `test/Sema/struct_differentiable.swift` to verify
generics don't trigger crashes. (Note that `Differentiable` derived conformances still
don't work with generics due to [SR-9595](https://bugs.swift.org/browse/SR-9595).)

dan-zheng committed:
[Sema] Fix `Differentiable` derived conformances crasher.

Previously, `getUnderlyingAllDiffableVariables` called `ProtocolConformance::getConcrete()`
unconditionally, assuming that the `ProtocolConformance` is concrete. However, the
conformance may be abstract (for generic type parameter types, for instance).

Check whether conformance is concrete before calling `getConcrete`. Tidy up related code.

Enable `-verify-ignore-unknown` for `test/Sema/struct_differentiable.swift` to verify
generics don't trigger crashes. (Note that `Differentiable` derived conformances still
don't work with generics due to SR-9595.)

dan-zheng commented:
@swift-ci Please test tensorflow

slavapestov commented on lib/Sema/DerivedConformanceAdditiveArithmeticVectorNumeric.cpp:
```diff
@@ -154,16 +154,16 @@ bool DerivedConformance::canDeriveAdditiveArithmetic(NominalTypeDecl *nominal,
+      return false;
+    auto declType = v->getType()->hasArchetype()
+        ? v->getType()
+        : DC->mapTypeIntoContext(v->getType());
```
mapTypeIntoContext() is a no-op when hasArchetype() is false

dan-zheng replied:
Thank you! Simplified in https://github.com/apple/swift/pull/22249/commits/e5d89ca2225027d68050e426ee1dc79b75bf881a.

slavapestov commented on lib/Sema/DerivedConformanceDifferentiable.cpp:
```diff
@@ -56,9 +56,9 @@ getStoredPropertiesForDifferentiation(NominalTypeDecl *nominal,
     if (vd->getAttrs().hasAttribute<NoDerivativeAttr>())
       continue;
-    if (!vd->hasInterfaceType())
+    if (!vd->hasType())
```
Why are you using the contextual type instead of the interface type everywhere?

dan-zheng replied:
I copied this logic from [`DerivedConformanceEquatableHashable.cpp`](https://github.com/apple/swift/blob/8928431cbee1739e9834a5cf0dcee8809278358b/lib/Sema/DerivedConformanceEquatableHashable.cpp#L54) (the upstream code has changed slightly, but still uses contextual types).

I believe contextual types are necessary because they store conformance info that isn't available in interface types.

Example:
```
// vd->getValueInterfaceType(), no conformances
(generic_type_param_type depth=0 index=0 decl=simple.(file).GenericVectorSpacesEqualSelf.T@simple.swift:31:37)

// vd->getType(), has conformances
(archetype_type name=T address=0x7fdda8856500 conforms_to=Swift.(file).AdditiveArithmetic conforms_to=Swift.(file).Differentiable
  0x7fdda504ec00 Module name=simple
    0x7fdda8800c08 FileUnit file="simple.swift"
      0x7fdda880f168 StructDecl name=GenericVectorSpacesEqualSelf

  (nested_type=AllDifferentiableVariables <<unresolved>>)
  (nested_type=CotangentVector =T)
  (nested_type=TangentVector =T))
```

dan-zheng replied:
On a related note, it definitely makes sense to move some duplicated helper `static` functions to a shared location (especially `allStoredPropertiesConformToProtocol`). [Here's a short list of functions that should be shared.](https://github.com/apple/swift/pull/21580#issuecomment-450713606)

@slavapestov: do you have any advice regarding this? My inclination is to move these functions to `CodeSynthesis.h`.

dan-zheng committed:
Minor edit.

Address feedback from @slavapestov.

dan-zheng commented:
Merging to unblock progress.
There's certainly room for improvement regarding derived conformances code, any feedback welcome!

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow linux GPU

dan-zheng committed:
Use interface type instead of contextual type.

Use `VarDecl::getValueInterfaceType()` where appropriate, mimicking code for
other derived conformances.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow
